README
#==============================================================
#
#  Name: README
#
#  Rem:  
#        EPICS IOC Application for LCLS-II superconducting LLRF
#
#  Side: None
#
#  Auth: 28-Jul-2016,          (GWBROWN)
#  Rev:  dd-mmm-yyyy, Reviewer's Name (USERNAME)
#--------------------------------------------------------------
#  Mod:
#        dd-mmm-yyyy, Reviewer's Name (USERNAME)
#          coment
#
#==============================================================
#

BUILDING: Requires code generated from register maps which are updated in scllrfApp/srcCheetah
To generate this code:
cd scllrfApp/srcCheetah
make
make copy

Then return to TOP (cd ../..) and make

This is the LLRF IOC for LLRF hardware using the network protocol from LBL for register
access. This is not part of the LCLS-II "Common Platform". The messages are formatted like:
64 bit nonce
32 bit function/address, 32 bit data
...
32 bit function/address, 32 bit data

The message must be smaller than the MTU for UDP, which is 1500 bytes, or about 175 registers.
The reply from the FPGA should be:
- The nonce echoed back unchanged
- For "read" requests, the function/address echoed back then requested 32 bit data
- for "write" requests, the function/adderss and data echoed back
(There is a bug in some firmware which adds a byte at the end of the response)

For simple cases, the code to read and write registers is generated by python Cheetah templates,
reading a dictionary/json file of register attributes.

LESS SIMPLE CASES:
CMOC fixed length waveforms are packed with the data for 8 waveforms interleaved.

32 bit function/address, 32 bit data for waveform 1
32 bit function/address, 32 bit data for waveform 2
...
32 bit function/address, 32 bit data for waveform 8
32 bit function/address, 32 bit data for waveform 1
32 bit function/address, 32 bit data for waveform 2
...

For PRC, there are up to 8 waveforms interleaved. The selection of waveforms
included are in the Trace Keep register. The trace status register bits 24 to
29 are the power of 2 for the array size, e.g. if they contain b000011, then
the waveform buffers are 2^3 = 8 registers long, or between one 8 point
waveform and 8 1 point waveforms.
Interleaving, if more than 1 waveform, is in the same manner as the CMOC waveforms.

# End of file


RELEASE_NOTES

#==============================================================
#
#  Name: RELEASE_NOTES
#
#  Rem:  This files describes release notes for
#        the gun and buncher LLRF IOC Application.
#
#  Auth: 28-Jul-2016, GARTH BROWN         (gwbrown)
#  Rev:  dd-mmm-yyyy, Reviewer's Name    (USERNAME)
#--------------------------------------------------------------

scllrf-R1-0-24: Soft trip on reflected power now uses average instead of peak power.

scllrf-R1-0-23: Added ASG of CAPUTLOG for all settings on screens.
				Added fault waveform saving for soft limit trips, more sophisticated turn-off.

scllrf-R1-0-22: A bunch of cleanup and refactoring for clarity.
				- Deleted experimental or obsolete files
				- Changed "extra" to dev, since it's the device layer, and added drv to driver layer things
				- This is not used for superconducting things, so scllrf -> gunbLlrf

scllrf-R1-0-21: Speed up frequency following loop. Make it triggered by new detune calculation.

scllrf-R1-0-20: newmat10 is not thread safe. Adding mutexes to make sure only one thread at a time uses it.
		Also adjusting exception handling to make sure bad data is not used or sent to PVs.

scllrf-R1-0-19: Added soft trip on high reflection level mid-pulse. Naming convention change.
				CATERs 1400006, 1400094, 141451, 141450,141473

scllrf-R1-0-18: Also migrating to EPICS 7
scllrf-R1-0-18-RC2: Hard coded which forward phase to use in CW detune. Experimental
				frequency correction. Phase mean calculation uses mean I and Q to avoid
				wrapping problems.
scllrf-R1-0-18-RC1: Reducing number of points for decay detune calc; buncher decays faster,
				was using noise for part of calculation. Test release.

scllrf-R1-0-17: Fixing "detune good" calculation. Use standard devitation of phase waveform,
				not last 300 phase means

scllrf-R1-0-16: Temperature tuning PID loop improved with PWM type control.

scllrf-R1-0-15: Added archive files.

scllrf-R1-0-14: Added buncher tuning PID loop, various small improvements based on ops

scllrf-R1-0-13: Adding a PV for loaded Q based on decay waveforms

scllrf-R1-0-12: Two new sequencers: Use "Extremum Seeking" to find minimum
				reflected RF, and monitor ADC board voltages

scllrf-R1-0-11: RF on/off behavior improved, consistent with SRF

scllrf-R1-0-10: Add power waveforms and average. Fix trip limit scaling. Phase balance seq.

scllrf-R1-0-9: Adding History module support for diagnostics and improved automation.

scllrf-R1-0-8: CW detune refinements.

scllrf-R1-0-7: Make CW detune work for buncher.

scllrf-R1-0-6: Fixing CW detune calculations.

scllrf-R1-0-5: Fixed status and processing settings in links.

scllrf-R1-0-4: Gun and buncher deployment

scllrf-R1-0-3: RFS waveforms scale with number of samples per point.
			Calculating average of amplitude and phase waveforms, and
			offering waveforms which average N previous waveforms.
			I/Q waveforms are now enabled in pairs, no reason to enable
			just one or the other.


gunbLlrfApp/src/BMB7Sup.c
/*
 * BPM device support
 */
#include <string.h>
#include <stdint.h> /* R3.14 version of epicsTypes.h has no 64 bit support */
#include <math.h>
#include <initHooks.h>
#include <epicsStdio.h>
#include <epicsString.h>
#include <epicsExit.h>
#include <epicsEvent.h>
#include <epicsThread.h>
#include <epicsTime.h>
#include <epicsMutex.h>
#include <epicsExport.h>
#include <cantProceed.h>
#include <alarm.h>
#include <iocsh.h>
#include <devLib.h>
#include <asynDriver.h>
#include <asynOctet.h>
#include <asynInt32.h>
#include <asynUInt32Digital.h>
#include <drvAsynIPPort.h>
#include <asynInt32Array.h>
#include <asynFloat32Array.h>
#include <asynFloat64Array.h>
#include <asynCommonSyncIO.h>

#include "BMB7Sup.h"

/*
 * ASYN subaddresses
 * Augment protocol command groups
 */
//#define A_HI_STATISTICS         0x0E00

/*
 * Number of times to retry a command
 */
#define COMMAND_RETRY_LIMIT 4

/*
 * Cell controller monitoring
 */
#define CC_MONITOR_UDP_PORT 50001
#define CC_MONITOR_REQUEST_REG_LIMIT     63
#define CC_MONITOR_RESPONSE_CAPACITY    187

/*
 * Links to lower port
 */
typedef struct portLink {
    char             *portName;
    char             *hostInfo;
    asynUser         *pasynUserCommon;
    asynUser         *pasynUserOctet;
    asynInterface    *poctet;
    int               isCommunicating;
} portLink;

/*
 * Driver private storage
 */
typedef struct drvPvt {
    /*
     * Asyn interfaces we provide
     */
    char                *portName;
    asynInterface        asynCommon;
    asynInterface        asynInt32;
    void                *asynInt32InterruptPvt;
    asynInterface        asynUInt32Digital;
    void                *asynUInt32DigitalInterruptPvt;

    /*
     * Link to lower ports
     */
    portLink             cmdLink;
    unsigned int         identifier;
    int                  cellInfo;
    portLink             monitorLink;


    /*
     * Statistics
     */
    unsigned long        commandCount[COMMAND_RETRY_LIMIT+1];
    unsigned long        commandFailedCount;

    /*
     * Monitoring
     */
    unsigned char         monitorRequest[2*CC_MONITOR_REQUEST_REG_LIMIT];
    unsigned char         monitorResponse[CC_MONITOR_RESPONSE_CAPACITY];
    int                   monitorIsActive;
    epicsTimeStamp        monitorWhenFailed;
} drvPvt;


static void
report(void *pvt, FILE *fp, int details)
{
    drvPvt *pdpvt = (drvPvt *)pvt;

    if (!pdpvt->cmdLink.isCommunicating)
        fprintf(fp, " Command UDP  not communicating\n");
    if (!pdpvt->monitorLink.isCommunicating)
        fprintf(fp, " Monitor UDP  not communicating\n");
}

static asynStatus
connect(void *pvt, asynUser *pasynUser)
{
    pasynManager->exceptionConnect(pasynUser);
    return asynSuccess;
}

static asynStatus
disconnect(void *pvt, asynUser *pasynUser)
{
    pasynManager->exceptionDisconnect(pasynUser);
    return asynSuccess;
}
static asynCommon commonMethods = { report, connect, disconnect };


/*
 * asynUInt32Digital methods -- there are none.
 * Everything is handled in SCAN="I/O Intr" callback.
 */
static asynUInt32Digital uint32DigitalMethods;

/*
 * asynInt32 methods
 */


static void
processBMB7MonitorResponse(drvPvt *pdpvt, int fault)
{
    ELLLIST *pclientList;
    interruptNode *pnode;
    extern volatile int interruptAccept;
    static const struct {
        unsigned char   hi;  /* Index into reply packet to hi byte of value */
        unsigned char   hiMask;      /* Valid bits in hi byte */
        unsigned char   signedWidth; /* 0 means unsigned */
    } addrMap[] = {
        { 138, 0x1F,  0 }, /*  0: LTC2990 A0 */
        { 136, 0x3F,  0 },
        { 134, 0x7F, 15 },
        {  58, 0x7F, 15 },
        { 132, 0x7F, 15 },
        {  56, 0x7F, 15 },

        { 130, 0x1F,  0 }, /*  6: LTC2990 A1 */
        { 128, 0x3F,  0 },
        { 126, 0x7F, 15 },
        {  54, 0x7F, 15 },
        { 124, 0x7F, 15 },
        {  52, 0x7F, 15 },

        { 122, 0x1F,  0 }, /* 12: LTC2990 A2 */
        { 120, 0x3F,  0 },
        { 118, 0x7F, 15 },
        {  50, 0x7F, 15 },

        { 114, 0x1F,  0 }, /* 16: LTC2990 B0 */
        { 112, 0x3F,  0 },
        { 110, 0x7F, 15 },
        {  46, 0x7F, 15 },
        { 108, 0x7F, 15 },
        {  44, 0x7F, 15 },

        { 106, 0x1F,  0 }, /* 22: LTC2990 B1 */
        { 104, 0x3F,  0 },
        { 102, 0x7F, 15 },
        {  42, 0x7F, 15 },
        { 100, 0x7F, 15 },
        {  40, 0x7F, 15 },

        {  98, 0x1F,  0 }, /* 28: LTC2990 B2 */
        {  96, 0x3F,  0 },
        {  94, 0x7F, 15 },
        {  38, 0x7F, 15 },
        {  92, 0x7F, 15 },
        {  36, 0x7F, 15 },

        {  90, 0x1F,  0 }, /* 34: LTC2990 C0 */
        {  88, 0x3F,  0 },
        {  86, 0x7F, 15 },
        {  34, 0x7F, 15 },
        {  84, 0x7F, 15 },
        {  32, 0x7F, 15 },

        {  82, 0x1F,  0 }, /* 40: LTC2990 C1 */
        {  80, 0x3F,  0 },
        {  78, 0x7F, 15 },
        {  30, 0x7F, 15 },
        {  76, 0x7F, 15 },
        {  28, 0x7F, 15 },

        {  74, 0x1F,  0 }, /* 46: LTC2990 C2 */
        {  72, 0x3F,  0 },
        {  70, 0x7F, 15 },
        {  26, 0x1F, 15 },
        {  68, 0x1F,  0 },

        {  66, 0x1F,  0 }, /* 51: LTC2990 C3 */
        {  64, 0x3F,  0 },
        {  62, 0x7F, 15 },
        {  22, 0x7F, 15 }
    };

    if (!interruptAccept) return;
    pasynManager->interruptStart(pdpvt->asynInt32InterruptPvt, &pclientList);
    pnode = (interruptNode *)ellFirst(pclientList);
    while (pnode) {
        asynInt32Interrupt *int32Interrupt = pnode->drvPvt;
        int idx = int32Interrupt->addr & ~CC_PROTOCOL_CMD_MASK_HI;
        pnode = (interruptNode *)ellNext(&pnode->node);
        if ((int32Interrupt->addr & CC_PROTOCOL_CMD_MASK_HI) ==
                                            CC_PROTOCOL_CMD_HI_BMB7_MONITOR) {
            epicsInt32 value;
            if (fault || (idx >= (sizeof addrMap / sizeof addrMap[0]))) {
                int32Interrupt->pasynUser->auxStatus = asynError;
                value = 0;
            }
            else {
                int h = CC_MONITOR_RESPONSE_CAPACITY-1-addrMap[idx].hi;
                int32Interrupt->pasynUser->auxStatus = asynSuccess;
                value = ((pdpvt->monitorResponse[h]&addrMap[idx].hiMask) << 8) |
                          pdpvt->monitorResponse[h+1];
                if (addrMap[idx].signedWidth) {
                    if (value >= (1 << (addrMap[idx].signedWidth - 1)))
                        value -= 1 << addrMap[idx].signedWidth;
                }
            }
            int32Interrupt->callback(int32Interrupt->userPvt,
                                     int32Interrupt->pasynUser,
                                     value);
        }
    }
    pasynManager->interruptEnd(pdpvt->asynInt32InterruptPvt);
}

/*
 * Set a mask/value request
 */
static void
setBMB7monitorRequest(drvPvt *pdpvt, unsigned int regIdx, int mask, int value)
{
    if (regIdx >= CC_MONITOR_REQUEST_REG_LIMIT) return;
    pdpvt->monitorRequest[1*CC_MONITOR_REQUEST_REG_LIMIT-1-regIdx] = mask;
    pdpvt->monitorRequest[2*CC_MONITOR_REQUEST_REG_LIMIT-1-regIdx] = value;
}

static asynStatus
writeReadBMB7monitor(asynUser *pasynUser, drvPvt *pdpvt)
{
    asynStatus status;
    asynOctet *pasynOctet = pdpvt->monitorLink.poctet->pinterface;
    size_t ntrans;
    int eomReason;

    pdpvt->monitorLink.pasynUserOctet->timeout = 0.2;
    pasynManager->lockPort(pdpvt->monitorLink.pasynUserOctet);
    status = pasynOctet->write(pdpvt->monitorLink.poctet->drvPvt,
                                      pdpvt->monitorLink.pasynUserOctet,
                                      (const char *)pdpvt->monitorRequest,
                                      2*CC_MONITOR_REQUEST_REG_LIMIT, &ntrans);
    if (status == asynSuccess) {
        status = pasynOctet->read(pdpvt->monitorLink.poctet->drvPvt,
                                      pdpvt->monitorLink.pasynUserOctet,
                                      (char *)pdpvt->monitorResponse,
                                      CC_MONITOR_RESPONSE_CAPACITY,
                                      &ntrans, &eomReason);
        asynPrint(pasynUser, ASYN_TRACEIO_DRIVER,
                    "Monitor read status %d, count %zu\n", (int)status, ntrans);
    }
    pasynManager->unlockPort(pdpvt->monitorLink.pasynUserOctet);
    if (status != asynSuccess) {
        epicsSnprintf(pasynUser->errorMessage, pasynUser->errorMessageSize,
                            "Monitor transfer failed: \"%s\"",
                            pdpvt->monitorLink.pasynUserOctet->errorMessage);
        if (pasynCommonSyncIO->disconnectDevice(pdpvt->monitorLink.pasynUserCommon) !=
                                                                    asynSuccess)
            asynPrint(pasynUser, ASYN_TRACE_ERROR, "%s can't disconnect: %s\n",
                            pdpvt->monitorLink.portName,
                            pdpvt->monitorLink.pasynUserCommon->errorMessage);
        processBMB7MonitorResponse(pdpvt, 1);
        epicsTimeGetCurrent(&pdpvt->monitorWhenFailed);
        pdpvt->monitorLink.isCommunicating = 0;
    }
    return status;
}

static asynStatus
crankBMB7monitor(asynUser *pasynUser, drvPvt *pdpvt)
{
    asynStatus status;
    epicsTimeStamp now;

    if (!pdpvt->monitorLink.isCommunicating) {
        epicsTimeGetCurrent(&now);
        if (epicsTimeDiffInSeconds(&now, &pdpvt->monitorWhenFailed) < 10)
            return asynError;
        status = pasynCommonSyncIO->connectDevice(pdpvt->monitorLink.pasynUserCommon);
        if (status != asynSuccess) {
            asynPrint(pasynUser, ASYN_TRACE_ERROR,
                        "%s failed to connect to %s: %s\n",
                            pdpvt->monitorLink.portName,
                            pdpvt->monitorLink.hostInfo,
                            pdpvt->monitorLink.pasynUserOctet->errorMessage);
            return status;
        }
        pdpvt->monitorLink.isCommunicating = 1;
        pdpvt->monitorIsActive = 0;
    }
    if (pdpvt->monitorIsActive) {
        if ((status = writeReadBMB7monitor(pasynUser, pdpvt)) != asynSuccess)
            return status;
        if (pdpvt->monitorResponse[CC_MONITOR_RESPONSE_CAPACITY-1-3] == 0x1) {
            processBMB7MonitorResponse(pdpvt, 0);
            pdpvt->monitorIsActive = 0;
        }
        else if (--pdpvt->monitorIsActive == 0) {
            processBMB7MonitorResponse(pdpvt, 1);
            return asynTimeout;
        }
    }
    else {
        setBMB7monitorRequest(pdpvt, 18, 0x10, 0x10);
        status = writeReadBMB7monitor(pasynUser, pdpvt);
        setBMB7monitorRequest(pdpvt, 18, 0x10, 0x00);
        if (status != asynSuccess)
            return status;
        if ((status = writeReadBMB7monitor(pasynUser, pdpvt)) != asynSuccess)
            return status;
        pdpvt->monitorIsActive = 10;
    }
    return asynSuccess;
}


static asynStatus
int32Read(void *pvt, asynUser *pasynUser, epicsInt32 *value)
{
    drvPvt *pdpvt = (drvPvt *)pvt;
    asynStatus status;
    int address;

    if ((status = pasynManager->getAddr(pasynUser, &address)) != asynSuccess)
        return status;

    status = crankBMB7monitor(pasynUser, pdpvt);
    return status;
}

static asynInt32 int32Methods = { NULL, int32Read };

/*
 * Create a new lower port and set up a link to it
 */
static asynStatus
setLink(drvPvt *pdpvt, const char *address, const char *ext, portLink *link, int udpPort, int priority)
{
    asynStatus status;

    link->portName = callocMustSucceed(1, strlen(pdpvt->portName)+strlen(ext)+1, "bmb7Conf");
    sprintf(link->portName, "%s%s", pdpvt->portName, ext);
    link->hostInfo = callocMustSucceed(1, strlen(address)+20, "bmb7Conf");
    sprintf(link->hostInfo, "%s:%d UDP", address, udpPort);
    /* No autoconnect, No process EOS */
    drvAsynIPPortConfigure(link->portName, link->hostInfo, priority, 1, 1);
    status = pasynCommonSyncIO->connect(link->portName, -1,
                                       &link->pasynUserCommon, NULL);
    if (status != asynSuccess) {
        printf("Can't set asynCommonSyncIO for port \"%s\".\n", link->portName);
        return asynError;
    }
    link->pasynUserOctet= pasynManager->createAsynUser(NULL, NULL);
    status = pasynManager->connectDevice(link->pasynUserOctet, link->portName, -1);
    if (status != asynSuccess) {
        printf("Can't find asyn port \"%s\".\n", link->portName);
        return asynError;
    }
    link->poctet = pasynManager->findInterface(link->pasynUserOctet, asynOctetType, 0);
    if (link->poctet == NULL) {
        printf("Can't find octet interface for \"%s\".\n", link->portName);
        return asynError;
    }
    link->pasynUserOctet->timeout = 5.0;
    return asynSuccess;
}

static void
bmb7Configure(const char *portName, const char *address,
               int priority)
{
    drvPvt *pdpvt;
    asynStatus status;
    int addressLen;

    /*
     * Handle defaults
     */
    if (priority == 0) priority = epicsThreadPriorityMedium;
    if ((portName == NULL) || (*portName == '\0')
     || (address == NULL)  || (*address == '\0')) {
        printf("Required argument not present!\n");
        return;
    }

    /*
     * Set up information for connection to BPM
     */
    if (strchr(address, ':') != NULL) {
        printf("Cell controller host info must not specify port.\n");
        return;
    }
    addressLen = strlen(address);
    if (address[addressLen - 1] == '*') {
        addressLen--;
        printf("Warning -- Broadcast designator ignored.\n");
    }

    /*
     * Set up local storage
     */
    pdpvt = (drvPvt *)callocMustSucceed(1, sizeof(drvPvt), "bmb7Conf");
    pdpvt->portName = epicsStrDup(portName);


    /*
     * Create the ports that we'll use to communicate with the BMB7.
     */
    if ((setLink(pdpvt, address, "_MON", &pdpvt->monitorLink,
                                CC_MONITOR_UDP_PORT, priority) != asynSuccess))
        return;

    /*
     * Create our port
     */
    status = pasynManager->registerPort(pdpvt->portName,
                                        ASYN_CANBLOCK|ASYN_MULTIDEVICE,
                                        1, 0, 0);
    if(status != asynSuccess) {
        printf("registerPort failed\n");
        return;
    }
    pdpvt->asynCommon.interfaceType = asynCommonType;
    pdpvt->asynCommon.pinterface  = &commonMethods;
    pdpvt->asynCommon.drvPvt = pdpvt;
    status = pasynManager->registerInterface(pdpvt->portName, &pdpvt->asynCommon);
    if (status != asynSuccess) {
        printf("registerInterface failed\n");
        return;
    }


    pdpvt->asynInt32.interfaceType = asynInt32Type;
    pdpvt->asynInt32.pinterface  = &int32Methods;
    pdpvt->asynInt32.drvPvt = pdpvt;
    status = pasynInt32Base->initialize(pdpvt->portName, &pdpvt->asynInt32);
    if (status != asynSuccess) {
        printf("pasynInt32Base->initialize failed\n");
        return;
    }
    pasynManager->registerInterruptSource(pdpvt->portName,
                                         &pdpvt->asynInt32,
                                         &pdpvt->asynInt32InterruptPvt);
    pdpvt->asynUInt32Digital.interfaceType = asynUInt32DigitalType;
    pdpvt->asynUInt32Digital.pinterface  = &uint32DigitalMethods;
    pdpvt->asynUInt32Digital.drvPvt = pdpvt;
    status = pasynUInt32DigitalBase->initialize(pdpvt->portName, &pdpvt->asynUInt32Digital);
    if (status != asynSuccess) {
        printf("pasynUInt32DigitalBase->initialize failed\n");
        return;
    }
    pasynManager->registerInterruptSource(pdpvt->portName,
                                         &pdpvt->asynUInt32Digital,
                                         &pdpvt->asynUInt32DigitalInterruptPvt);

}

/*
 * IOC shell command registration
 */
static const iocshArg bmb7ConfigureArg0 ={ "port",             iocshArgString};
static const iocshArg bmb7ConfigureArg1 ={ "address",          iocshArgString};
static const iocshArg bmb7ConfigureArg2 ={ "priority",         iocshArgInt};
static const iocshArg *bmb7ConfigureArgs[] = {
                    &bmb7ConfigureArg0, &bmb7ConfigureArg1,
                    &bmb7ConfigureArg2};
static const iocshFuncDef bmb7ConfigureFuncDef =
                                     {"bmb7Configure", 3, bmb7ConfigureArgs};
static void bmb7ConfigureCallFunc(const iocshArgBuf *args)
{
    bmb7Configure(args[0].sval, args[1].sval,
                   args[2].ival);
}

static void
bmb7_RegisterCommands(void)
{
    iocshRegister(&bmb7ConfigureFuncDef, bmb7ConfigureCallFunc);
}
epicsExportRegistrar(bmb7_RegisterCommands);
